evals/funnels/sql/steps.py

```diff
@@ -0,0 +1,222 @@
+import pandas as pd
+from dataclasses import dataclass
+from typing import List, Dict, Any, Optional, Tuple
+from evals.eval_types import TableDetails, SQLDetails
+from evals.funnels.sql.utils import parse_table_path, run_sql_query
+
+
+@dataclass
+class FunnelStepResult:
+    name: str
+    passed: bool
+    notes: Optional[str] = None
+
+
+def sql_generated_test(
+    expected_output: str | None,
+    sql_query_recieved: str | None,
+) -> FunnelStepResult:
+    """
+    Verifies whether a SQL query was generated as expected.
+
+    This is the first funnel in the SQL evaluation process. It checks if the AI generated
+    a SQL query when it was supposed to, or didn't generate one when it wasn't supposed to.
+
+    Args:
+        expected_output: The expected output string or None if no SQL query was expected
+        sql_query_recieved: The actual SQL query string generated or None if no query was generated
+    """
+    name = "sql_generated_test"
+
+    if (expected_output is None) and (sql_query_recieved is None):
+        # No SQL query was expected and none was generated
+        return FunnelStepResult(name=name, passed=True)
+    elif (expected_output is not None) and (sql_query_recieved is not None):
```

Comment:
Maybe just so we are not so fragile to AI to return None when generating the SQLDetails, we could do:

```python
expected_sql_query_exists = expected_output is not None and expected_output != ''
actual_sql_query_exists = sql_query_recieved is not None and sql_query_recieved != ''

...
```